### ISSUE TITLE
`Self` is incompatible with instance constructed from partial of `cls` parameter

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#6378]
**Describe the bug**
When passing a `type[Self]` to a `partial`, the type parameter is concretised and cannot be returned as `Self`.

**Code or Screenshots**

```python
# pyright: strict

from dataclasses import dataclass
from functools import partial
from typing import Self, reveal_type

@dataclass
class Foo:
    bar: int
    baz: str

    @classmethod
    def from_bar(cls, bar: int) -> Self:
        reveal_type(cls)  # type[Self@Foo]
        make_foo = partial(cls, bar=bar)
        reveal_type(make_foo)  # partial[Foo]
        self = make_foo(baz='')
        reveal_type(self)  # Foo
        other_self = cls(bar=1, baz='')
        reveal_type(other_self)  # Self@Foo
        return self  # Error: Expression of type "Foo" cannot be assigned to return type "Self@Foo"


```

**VS Code extension or command-line**
Both, v1.1.335

**Comment by layday** [comment]
More examples if they help: https://github.com/layday/instawow/actions/runs/6794663050/job/18471457823. Also affects enum members when iterating over an enum of type `Self`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

If you'd like to avoid being impacted by future regressions, you could add your project to [mypy_primer](https://github.com/hauntsaninja/mypy_primer/blob/f02e889c8313b87df09578c7a3ce8af74a26136c/mypy_primer/projects.py#L61). This tool is used for regression testing on pyright, mypy, and typeshed changes.

**Comment by layday** [comment]
Thank you, I'll look into that.

**Comment by NeilGirdhar** [comment]
I guess this is the same bug:
```python
import dataclasses
from dataclasses import dataclass, replace
from typing import Any, ClassVar, Generic, Protocol, Self, TypeVar


class DataclassInstance(Protocol):
    __dataclass_fields__: ClassVar[dict[str, dataclasses.Field[Any]]]


T = TypeVar('T')
@dataclass
class X(Generic[T]):
    a: int

    def f(self) -> Self:
        reveal_type(self)  # Self@X[T@X]
        assert isinstance(self, DataclassInstance)
        reveal_type(self)  # X[T@X]
        return replace(self, a=1)  # Error, can't return X.
```

**Comment by erictraut** [comment]
@NeilGirdhar, no that's not the same issue.

**Comment by NeilGirdhar** [comment]
@erictraut Should I file a separate issue?

**Comment by erictraut** [comment]
Yes, please

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.336, which I just published. It will also be included in a future release of pylance.
